MelodyGenerator.js

```diff
@@ -0,0 +1,145 @@
+		
+
+		// [2] contains 'pause' or 'note'
+		// [3] is the note letter(s) or duration of the pause
+		// [8] is the duration of a note command
+		
+		if(regex1.test(soundString)) { // In case of repeat commands, we'll need a loop
+			//extract repeat command: get times, and note/macro
+			let noteArray = [];
+			const repeat = regex1.exec(soundString);
+			// [1] is number of times
+			// [2] is the note or macro
+
+			
+			const repeatSound = this.parseSound(repeat[2].trim());  //RECURSION!
+			// loop X times through the given note or macro
+			// console.log(repeat, macros.get(repeat));
+
+			for (let i = 0; i < parseInt(repeat[1]); i++) {
+				noteArray = [...noteArray, ...repeatSound]; 
+			}
+
+			return noteArray;
+
+		} else if (soundTest && sound[2] === 'note(') { //it must be just a plain old note
+			let noteArray = sound[3].split(";");
+			// let noteString = "durations[" + sound[8] + "],";
+
+			// for(let i=0; i < noteArray.length; i++){
+			// 	noteString += " note('" + noteArray[i] + "'),";
+			// }
+
+			// noteString = noteString.slice(0, -1);
+			// console.log(noteString);
+			// melody.enqueueTone(...noteString);
+			// console.log(noteArray);
+
+			return [[sound[8], ...noteArray]];
+
+		} else if (soundTest && sound[2] === 'pause(') { // if it's a pause, then... pause, obviously
+			return [[sound[3]]];
+
+		} else if (macros.has(regex4.exec(soundString)[2].trim())) { // It must be a macro
+
+			return macros.get(regex4.exec(soundString)[2].trim()); 
+		}
+	}
+
+	parseLine(lineString) {
+
+		if(regex1.test(lineString) || regex2.test(lineString) || regex3.test(lineString) || regex4.test(lineString)) {
+
+			let notesArray = [];
+
+			if(lineString.startsWith('define')) {
+					this.parseMacro(lineString); // Adds the macro to the map
+					return null;
+			} else {
+				let line = lineString.split("|"); //split line by vertical bars into individual sounds
+
+				for(let i=0; i < line.length; i++) {
+					// Extract the sound out of the array
+					// notesArray.push(this.parseSound(line[i]));
+					// console.log(line[i]);
+					notesArray = [...notesArray, ...this.parseSound(line[i].trim())];
+				}
+			return notesArray;	
+			}
+		}
+
+		return null;		
+	}
+
+    fromString(melodyString) {
+
+        //  Split by lines, trim
+        const melodyLines = melodyString.trim().split("\n");
+
+        // Loop thru all the lines
+        for(let i=0; i < melodyLines.length; i++) {
+        	// Extract an array of just notes or pauses
+        	let notesArray = this.parseLine(melodyLines[i]);
+        	if(notesArray !== null && notesArray !== undefined && notesArray[0] !== undefined) {
+	        	// console.log(notesArray);
+	        	//Add each note to the queue
+	        	for(let j=0; j < notesArray.length; j++){
+	        		if(notesArray[j].length === 1) {
+	        			melody.enqueuePause(durations[notesArray[j][0]]);
+	        			// console.log('pause');
+	        		} else if(notesArray[j].length === 2) {
```

Comment:
is not a good option to detect type of structure with length array. What if I would create macros with arguments? 
I preffer to 'eval' send special object with type and named variables. Now I must to guess arguments and what is a object

Reply:
I think the code could easily be adapted to macros with arguments. Macros have to be defined and the code is set up to catch them.

My detection method is based on the patterns of the `notes` method, i.e. the first argument passed is duration, all the rest are tones. If there's not tone and only duration, it has to be a pause, right? 

> I preffer to 'eval' send special object with type and named variables. Now I must to guess arguments and what is a object
I'm not sure what you mean, can you write it out or send me a link?